### PR TITLE
Push generic stuff to the JDBC level

### PR DESCRIPTION
**Pull request by tirsen**
This separates generic database interaction "stuff" like connection
pooling from Hibernate setup.

This accomplishes two things:
* Opens up for using other database interaction frameworks in the
  future.
* Forces us to build scalability sanity checks in such a way that they
  do not overuse Hibernate trickeries.

This is the first step towards extracting all this to a misk-jdbc module
which does not depend on misk-hibernate and should be able to stand on
its own.

Also includes a cross shard query detector as an example of what we can
build at this level.

### CODE REVIEW

File: misk-hibernate/src/main/kotlin/misk/jdbc/DataSourceService.kt [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import misk.environment.Environment
+import misk.inject.toKey
+import mu.KotlinLogging
+import javax.inject.Provider
+import javax.inject.Singleton
+import javax.sql.DataSource
+import kotlin.reflect.KClass
+
+/**
+ * Builds a bare connection to a Hibernate database. Doesn't do any schema migration or validation.
```
Comment (ryanhall07) [inline misk-hibernate/src/main/kotlin/misk/jdbc/DataSourceService.kt]: this probably shouldn't reference Hibernate

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+import net.ttddyy.dsproxy.support.ProxyDataSource
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okio.BufferedSource
+import okio.ByteString.Companion.encodeUtf8
+import java.io.EOFException
+import javax.inject.Singleton
+import javax.sql.DataSource
+
+data class Instruction(
```
Comment (ryanhall07) [inline misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt]: should these data classes be private? or at least internal?

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  val Instructions: Instruction,
+  val ExecCount: Int
+) {
+  val isScatter: Boolean
+    get() = Instructions.isScatter
+}
+
+/**
+ * Throws a [CrossShardQueryException] for scatter queries that doesn't have a lookup vindex.
+ * Note: Current implementation is not thread safe and will not work in production.
```
Comment (swankjesse) [inline misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt]: this is an interesting comment . . . a non-threadsafe singleton makes me itchy

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+class CrossShardQueryDetector(
+  val okHttpClient: OkHttpClient,
+  val moshi: Moshi,
+  val config: DataSourceConfig
+) : DataSourceDecorator {
+
+  private val listener = Listener()
+
+  override fun decorate(dataSource: DataSource): DataSource {
+    if (config.type != DataSourceType.VITESS) return dataSource
```
Comment (swankjesse) [inline misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt]: should this be called out in the API? Maybe the classname?

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+      count.set(extractScatterQueryCount())
+    }
+
+    override fun afterQuery(execInfo: ExecutionInfo, queryInfoList: MutableList<QueryInfo>) {
+      if (disabled.get()) return
+
+      val newScatterQueryCount = extractScatterQueryCount()
+      if (newScatterQueryCount > count.get()) {
+        throw CrossShardQueryException()
```
Comment (swankjesse) [inline misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt]: we’ll probably want more context once we start using this!

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+      } catch (e: EOFException) {
+        null
+      }
+    }.filterNotNull()
+  }
+
+  fun <T> disable(body: () -> T): T  = listener.disable(body)
+}
+
+private inline fun <T, R> ThreadLocal<T>.withValue(value: T, body: () -> R): R {
```
Comment (swankjesse) [inline misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt]: coool

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/StartVitessService.kt [hunk 7]
```diff
@@ -1,4 +1,4 @@
-package misk.hibernate
+package misk.jdbc
 
 import com.github.dockerjava.api.DockerClient
 import com.github.dockerjava.api.model.*
```
Comment (swankjesse) [inline misk-hibernate-testing/src/main/kotlin/misk/jdbc/StartVitessService.kt]: nit: no wildcards!

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/StartVitessService.kt [hunk 8]
```diff
@@ -136,7 +134,9 @@ internal class StartVitessService(val config: DataSourceConfig) : AbstractIdleSe
     val clusters = CacheBuilder.newBuilder()
         .removalListener<DataSourceConfig, DockerVitessCluster> { it.value.stop() }
         .build(CacheLoader.from { config: DataSourceConfig? ->
-          DockerVitessCluster(config!!, docker, gson)
+          DockerVitessCluster(config!!,
+              docker,
+              gson)
```
Comment (swankjesse) [inline misk-hibernate-testing/src/main/kotlin/misk/jdbc/StartVitessService.kt]: nit: ugh gson!

File: misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt [hunk 9]
```diff
@@ -4,6 +4,10 @@ import com.google.common.collect.ImmutableSet
 import io.opentracing.Tracer
 import io.opentracing.tag.Tags
 import misk.backoff.ExponentialBackoff
+import misk.jdbc.DataSourceConfig
```
Comment (swankjesse) [inline misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt]: unused?

File: misk-hibernate/src/test/kotlin/misk/hibernate/ScaleSafetyTest.kt [hunk 10]
```diff
@@ -37,35 +40,109 @@ class ScaleSafetyTest {
...
-//    assertThrows<CrossShardTransactionException> {
+    assertThrows<CrossShardTransactionException> {
       transacter.transaction { session ->
         session.save(DbCharacter("Ian Malcolm", session.load(jp), session.load(jg)))
         session.save(DbCharacter("Leia Organa", session.load(sw), session.load(cf)))
       }
-//    }
-//    assertThrows<CrossShardTransactionException> {
+    }
+    assertThrows<CrossShardTransactionException> {
```
Comment (swankjesse) [inline misk-hibernate/src/test/kotlin/misk/hibernate/ScaleSafetyTest.kt]: sweet

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt (hunk 3 above)
Comment (tirsen) [reply misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt]: Yeah @systay gave me shit about this too. I think I'm gonna have to rethink how I do this. Currently this is only used in test of course. Do you mind if I mull this over for a bit and return to this in a follow up PR?

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt (hunk 4 above)
Comment (tirsen) [reply misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt]: I renamed it to "VitessScatterException"

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt (hunk 6 above)
Comment (tirsen) [reply misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt]: I love kotlin!

File: misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt (hunk 9 above)
Comment (tirsen) [reply misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt]: it's used

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+          throw CrossShardQueryException()
+        }
+      }
+    })
+    return proxy
+  }
+
+  private fun extractScatterQueryCount(): Int {
+    val request = Request.Builder()
+        .url("http://localhost:27000/debug/query_plans")
```
[the remark this answers is not in the document]
Comment (tirsen) [reply misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt]: maybe but we'll notice

**Pull request merged**